### ISSUE TITLE
`TrainingConfigurator` の責務分離

### ISSUE THREAD

**Issue by kurorosu** [kurorosu/pochitrain#315]
## Summary

`TrainingConfigurator` が optimizer/scheduler/criterion 構築に加えて, 層別学習率の処理 (パラメータグループ構築, ResNet層名抽出, ロギング) を兼務しており, SRP に違反している. 層別学習率ロジックを Strategy パターンで分離する.

## Background

- `training_configurator.py` (269行) の `configure()` メソッドが 5つの責務を統合している
- `_get_layer_group()` が ResNet 構造にハードコーディングされており, 他モデル (EfficientNet/MobileNet) への拡張が不可能
- 層別学習率ロジックが TrainingConfigurator 内に密結合しているため, 単体テストが困難
- `optimization/param_suggestor.py` では既に Strategy パターンが採用されており, 同じアプローチで統一できる

### 1. TrainingConfigurator の責務一覧

| メソッド | 行数 | 責務 | SRP違反 |
|---------|------|------|---------|
| `configure()` | 57 | 5つの構築処理のオーケストレーション + ロギング | 中 |
| `_build_criterion()` | 26 | 損失関数構築 + バリデーション + デバイス転送 | 低 |
| `_build_optimizer()` | 25 | optimizer 構築 (名前→クラスマッピング) | 低 |
| `_build_scheduler()` | 40 | scheduler 構築 (名前→クラスマッピング) | 低 |
| `_build_layer_wise_param_groups()` | 38 | 層別パラメータグループ構築 | **高** |
| `_get_layer_group()` | 26 | パラメータ名→層グループ名抽出 (ResNet専用) | **高** |
| `_log_layer_wise_lr()` | 13 | 層別学習率のログ出力 | 中 |

### 2. `_get_layer_group()` の ResNet ハードコーディング

```python
def _get_layer_group(self, param_name: str) -> str:
    if "layer1" in param_name:
        return "layer1"
    elif "layer2" in param_name:
        return "layer2"
    elif "layer3" in param_name:
        return "layer3"
    elif "layer4" in param_name:
        return "layer4"
    elif "conv1" in param_name:
        return "conv1"
    elif "bn1" in param_name:
        return "bn1"
    elif "fc" in param_name:
        return "fc"
    else:
        return "other"
```

- ResNet 以外のモデルでは **全パラメータが "other" に分類される**
- EfficientNet (`_blocks[0]._expand_conv.weight`) や MobileNet (`features.0.0.weight`) では層別学習率が機能しない
- Issue #323 (EfficientNet/MobileNet 対応) の前提条件となる

### 3. テストカバレッジの状況

| テスト対象 | カバレッジ | 課題 |
|-----------|----------|------|
| `configure()` 公開API | 100% | 良好 |
| `_build_optimizer/scheduler/criterion` | 100% | 良好 |
| `_build_layer_wise_param_groups()` | 70-80% | requires_grad=False の edge case 未テスト |
| `_get_layer_group()` | 部分的 | 7分岐中3つのみ間接テスト (layer2,3,4, bn1, other 未テスト) |
| `_log_layer_wise_lr()` | 0% | テストなし |

### 4. データフロー

```
PochiTrainer.__init__()
  └─ TrainingConfigurator(device, logger)

PochiTrainer.setup_training()
  └─ TrainingConfigurator.configure()
       ├─ _build_criterion()
       ├─ _build_layer_wise_param_groups()  ← 分離対象
       │    └─ _get_layer_group()           ← ResNet専用
       ├─ _log_layer_wise_lr()              ← 分離対象
       ├─ _build_optimizer()
       └─ _build_scheduler()
            ↓
       TrainingComponents (dataclass)
            ↓
       PochiTrainer のインスタンス変数に展開
            ↓
       TrainingLoop → MetricsTracker (層別学習率の可視化)
```

### Strategy パターンによる層別学習率の分離

`optimization/param_suggestor.py` の設計に倣い, 層別学習率ロジックを Strategy パターンで分離する.

#### 新規ファイル構成

```
pochitrain/training/
  layer_wise_lr/
    __init__.py
    interfaces.py              # ILayerGrouper インターフェース
    resnet_grouper.py          # ResNet用の層グループ抽出 (現在の _get_layer_group)
    param_group_builder.py     # パラメータグループ構築 (現在の _build_layer_wise_param_groups)
```

#### インターフェース定義

```python
# pochitrain/training/layer_wise_lr/interfaces.py
from abc import ABC, abstractmethod

class ILayerGrouper(ABC):
    """パラメータ名から層グループ名を抽出するインターフェース."""

    @abstractmethod
    def get_group(self, param_name: str) -> str:
        """パラメータ名から層グループ名を返す."""
```

#### ResNet 実装

```python
# pochitrain/training/layer_wise_lr/resnet_grouper.py
class ResNetLayerGrouper(ILayerGrouper):
    """ResNet構造に基づく層グループ抽出."""

    def get_group(self, param_name: str) -> str:
        # 現在の _get_layer_group() のロジックをそのまま移動
```

#### TrainingConfigurator の変更

```python
class TrainingConfigurator:
    def __init__(
        self,
        device: torch.device,
        logger: logging.Logger,
        layer_grouper: Optional[ILayerGrouper] = None,  # DIP: 依存性注入
    ):
        self.device = device
        self.logger = logger
        self.layer_grouper = layer_grouper
```

- `_get_layer_group()` と `_build_layer_wise_param_groups()` を削除
- `_log_layer_wise_lr()` を `param_group_builder.py` に移動
- `configure()` 内の層別学習率分岐は `param_group_builder.build()` に委譲

#### PochiTrainer での注入

```python
# pochi_trainer.py
from pochitrain.training.layer_wise_lr import ResNetLayerGrouper

class PochiTrainer:
    def __init__(self, ...):
        grouper = ResNetLayerGrouper()
        self.training_configurator = TrainingConfigurator(
            self.device, self.logger, layer_grouper=grouper
        )
```

### 将来の拡張性

Issue #323 (EfficientNet/MobileNet 対応) 時に, 新しい Grouper を追加するだけで対応可能:

```python
class EfficientNetLayerGrouper(ILayerGrouper):
    def get_group(self, param_name: str) -> str:
        if "_blocks" in param_name:
            # ブロックインデックスに基づくグループ化
            ...
        elif "_fc" in param_name:
            return "head"
        ...
```

## Acceptance Criteria

- 層別学習率ロジックが `training/layer_wise_lr/` パッケージに分離されていること
- `ILayerGrouper` インターフェースが定義され, `ResNetLayerGrouper` が実装されていること
- `TrainingConfigurator` から `_get_layer_group()`, `_build_layer_wise_param_groups()`, `_log_layer_wise_lr()` が除去されていること
- `TrainingConfigurator` が `ILayerGrouper` を依存性注入で受け取ること
- `ResNetLayerGrouper` の全分岐 (layer1-4, conv1, bn1, fc, other) の単体テストがあること
- 既存テストがすべてパスすること
- 訓練結果 (層別学習率の動作) に変更がないこと

## Labels

`refactor`

## Branch

`refactor/training-configurator-srp`
